File: apps/android/app/src/main/java/com/sanvya/app/ui/dashboard/AddWidgetSheet.kt
package com.sanvya.app.ui.dashboard

import androidx.compose.foundation.layout.Arrangement
import androidx.compose.foundation.layout.Column
import androidx.compose.foundation.layout.Row
import androidx.compose.foundation.layout.fillMaxWidth
import androidx.compose.foundation.layout.padding
import androidx.compose.foundation.rememberScrollState
import androidx.compose.foundation.verticalScroll
import androidx.compose.runtime.Composable
import androidx.compose.runtime.collectAsState
import androidx.compose.runtime.getValue
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.unit.dp
import com.sanvya.app.i18n.S
import com.sanvya.app.i18n.sRes
import com.sanvya.app.theme.LocalSanvyaColors
import com.sanvya.app.theme.SanvyaType
import com.sanvya.app.ui.components.SanvyaButton
import com.sanvya.app.ui.components.SanvyaCard
import com.sanvya.app.ui.components.SanvyaModal
import com.sanvya.app.ui.components.SanvyaText

/**
 * Web's "Add a widget" modal.
 *
 * **It only offers tiles that are built.** Web can list all fourteen because
 * all fourteen render; here `TileId.isBuilt` gates the list, so a user can
 * never add a tile that would appear as an empty card. The eleven unbuilt ones
 * are absent from the picker, not present-and-broken, and are tracked in
 * ABSENT-BY-DECISION.md.
 *
 * A premium tile the user is not entitled to is shown **locked rather than
 * hidden**, which is web's behaviour: knowing the feature exists is the point
 * of the lock. It is only ever a locked row here, never a locked tile on the
 * dashboard — those are filtered out of the grid entirely.
 */
@Composable
fun AddWidgetSheet(
    open: Boolean,
    isPaid: Boolean,
    onClose: () -> Unit,
) {
    val colors = LocalSanvyaColors.current
    val enabled by DashboardPrefs.ids.collectAsState()
    val available = TileId.entries.filter { it.isBuilt && it !in enabled }

    SanvyaModal(open = open, onClose = onClose, label = S.Dashboard.addWidget(sRes())) {
        Column(verticalArrangement = Arrangement.spacedBy(4.dp), modifier = Modifier.padding(bottom = 12.dp)) {
            SanvyaText(S.Dashboard.addWidget(sRes()), style = SanvyaType.h2)
            SanvyaText(
                S.Dashboard.addWidgetIntro(sRes()),
                style = SanvyaType.statLabel,
                color = colors.text2,
            )
        }

        if (available.isEmpty()) {
            SanvyaText(S.Dashboard.allAdded(sRes()), style = SanvyaType.statLabel, color = colors.text2)
        } else {
            Column(
                verticalArrangement = Arrangement.spacedBy(8.dp),
                modifier = Modifier.verticalScroll(rememberScrollState()),
            ) {
                available.forEach { tile ->
                    val locked = tile.isPremium && !isPaid
                    SanvyaCard(modifier = Modifier.fillMaxWidth()) {
                        Row(
                            verticalAlignment = Alignment.CenterVertically,
                            modifier = Modifier.fillMaxWidth(),
                        ) {
                            Column(Modifier.weight(1f)) {
                                SanvyaText(
                                    tile.title(sRes()),
                                    style = SanvyaType.body,
                                    color = if (locked) colors.text2 else colors.text,
                                )
                                if (locked) {
                                    SanvyaText(
                                        S.Dashboard.premium(sRes()),
                                        style = SanvyaType.statLabel,
                                        color = colors.text2,
                                    )
                                }
                            }
                            SanvyaButton(
                                onClick = { DashboardPrefs.setEnabled(tile, true) },
                                enabled = !locked,
                            ) {
                                SanvyaText(S.Translation.commonAdd(sRes()), style = SanvyaType.button)
                            }
                        }
                    }
                }
            }
        }

        if (!isPaid) {
            SanvyaText(
                S.Dashboard.premiumNote(sRes()),
                style = SanvyaType.statLabel,
                color = colors.text2,
                modifier = Modifier.padding(top = 10.dp),
            )
        }

        Row(modifier = Modifier.fillMaxWidth().padding(top = 14.dp), horizontalArrangement = Arrangement.End) {
            SanvyaButton(onClick = onClose) {
                SanvyaText(S.Translation.commonDone(sRes()), style = SanvyaType.button)
            }
        }
    }
}

File: apps/ios/App/AddWidgetSheet.swift
import SwiftUI

/**
 Web's "Add a widget" modal.

 **It only offers tiles that are built.** Web can list all fourteen because all
 fourteen render; here `TileId.isBuilt` gates the list, so a user can never add
 a tile that would appear as an empty card. The eleven unbuilt ones are absent
 from the picker, not present-and-broken, and are tracked in
 ABSENT-BY-DECISION.md.

 A premium tile the user is not entitled to is shown **locked rather than
 hidden**, which is web's behaviour: knowing the feature exists is the point of
 the lock. It is only ever a locked row here, never a locked tile on the
 dashboard — those are filtered out of the grid entirely.

 Mirrors `apps/android/.../ui/dashboard/AddWidgetSheet.kt`.
 */
struct AddWidgetSheet: View {
    let isPaid: Bool
    let onClose: () -> Void

    @ObservedObject private var prefs = DashboardPrefs.shared

    private var available: [TileId] {
        TileId.allCases.filter { $0.isBuilt && !prefs.ids.contains($0) }
    }

    var body: some View {
        VStack(alignment: .leading, spacing: 0) {
            VStack(alignment: .leading, spacing: 4) {
                SanvyaH2(S.Dashboard.addWidget)
                Text(S.Dashboard.addWidgetIntro)
                    .sanvyaStyle(SanvyaType.statLabel)
                    .foregroundStyle(Color.text2)
                    .fixedSize(horizontal: false, vertical: true)
            }
            .padding(.bottom, 12)

            if available.isEmpty {
                Text(S.Dashboard.allAdded)
                    .sanvyaStyle(SanvyaType.statLabel)
                    .foregroundStyle(Color.text2)
            } else {
                ScrollView {
                    VStack(spacing: 8) {
                        ForEach(available) { tile in
                            row(tile)
                        }
                    }
                }
                .frame(maxHeight: 360)
            }

            if !isPaid {
                Text(S.Dashboard.premiumNote)
                    .sanvyaStyle(SanvyaType.statLabel)
                    .foregroundStyle(Color.text2)
                    .fixedSize(horizontal: false, vertical: true)
                    .padding(.top, 10)
            }

            HStack {
                Spacer(minLength: 0)
                SanvyaButton(action: onClose) { Text(S.Translation.commonDone) }
            }
            .padding(.top, 14)
        }
    }

    private func row(_ tile: TileId) -> some View {
        let locked = tile.isPremium && !isPaid
        return SanvyaCard {
            HStack(spacing: 10) {
                VStack(alignment: .leading, spacing: 2) {
                    Text(tile.title)
                        .sanvyaStyle(SanvyaType.body)
                        .foregroundStyle(locked ? Color.text2 : Color.text)
                    if locked {
                        Text(S.Dashboard.premium)
                            .sanvyaStyle(SanvyaType.statLabel)
                            .foregroundStyle(Color.text2)
                    }
                }
                Spacer(minLength: 0)
                SanvyaButton { prefs.setEnabled(tile, true) } label: {
                    Text(S.Translation.commonAdd)
                }
                .disabled(locked)
            }
            .frame(maxWidth: .infinity, alignment: .leading)
        }
    }
}
